File: app/src/main/java/com/example/telemedicine_lab5/ui/screens/CallScreen.kt
package com.example.telemedicine_lab5.ui.screens

import android.Manifest
import android.util.Log
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CallEnd
import androidx.compose.material.icons.filled.Cameraswitch
import androidx.compose.material.icons.filled.Mic
import androidx.compose.material.icons.filled.MicOff
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.FloatingActionButton
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.compose.ui.viewinterop.AndroidView
import androidx.navigation.NavController
import com.example.telemedicine_lab5.signaling.SignalingMessage
import com.example.telemedicine_lab5.signaling.SocketSignalingClient
import com.example.telemedicine_lab5.webrtc.WebRtcSessionManager
import kotlinx.coroutines.launch
import org.webrtc.IceCandidate
import org.webrtc.PeerConnection
import org.webrtc.SurfaceViewRenderer

private const val TAG = "CallScreen"

@Composable
fun CallScreen(
    navController: NavController,
    role: String,
    roomId: String,
    serverUrl: String,
) {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()

    val signalingClient = remember { SocketSignalingClient() }
    val sessionManager = remember { WebRtcSessionManager(context.applicationContext) }

    var localRenderer by remember { mutableStateOf<SurfaceViewRenderer?>(null) }
    var remoteRenderer by remember { mutableStateOf<SurfaceViewRenderer?>(null) }

    var status by remember { mutableStateOf("Requesting camera and microphone permissions...") }
    var isMuted by remember { mutableStateOf(false) }
    var isConnected by remember { mutableStateOf(false) }
    var hasRemoteVideo by remember { mutableStateOf(false) }
    var started by remember { mutableStateOf(false) }
    var offerSent by remember { mutableStateOf(false) }
    var isOfferer by remember { mutableStateOf(false) }
    var isRecoveringIce by remember { mutableStateOf(false) }
    var hasRequestedPermissions by remember { mutableStateOf(false) }

    val permissionsLauncher = rememberLauncherForActivityResult(
        contract = ActivityResultContracts.RequestMultiplePermissions(),
    ) { permissions ->
        val granted = permissions[Manifest.permission.CAMERA] == true &&
            permissions[Manifest.permission.RECORD_AUDIO] == true

        if (!granted) {
            status = "Camera and microphone permissions are required"
            return@rememberLauncherForActivityResult
        }

        scope.launch {
            runCatching {
                startSession(
                    roomId = roomId,
                    serverUrl = serverUrl,
                    signalingClient = signalingClient,
                    sessionManager = sessionManager,
                    onStatusChange = { status = it },
                    onConnected = { isConnected = it },
                    onRemoteVideo = { hasRemoteVideo = true },
                    onIncomingMessage = { message ->
                        scope.launch {
                            handleSignalingMessage(message, roomId, signalingClient, sessionManager, onStatusChange = { status = it })
                        }
                    },
                    onReadyToOffer = {
                        scope.launch {
                            if (offerSent) return@launch
                            runCatching {
                                isOfferer = true
                                offerSent = true
                                status = "Creating offer..."
                                val offer = sessionManager.createOffer()
                                signalingClient.sendOffer(roomId, offer.description)
                                status = "Offer sent, waiting for answer"
                            }.onFailure {
                                offerSent = false
                                status = "Offer error: ${it.message}"
                            }
                        }
                    },
                    onIceFailed = {
                        if (!started || isRecoveringIce) {
                            Unit
                        } else if (!isOfferer) {
                            status = "Peer reconnection in progress..."
                        } else {
                            scope.launch {
                                isRecoveringIce = true
                                runCatching {
                                    status = "Connection failed, restarting ICE..."
                                    val restartOffer = sessionManager.createIceRestartOffer()
                                    signalingClient.sendOffer(roomId, restartOffer.description)
                                    status = "ICE restart offer sent"
                                }.onFailure {
                                    status = "ICE restart error: ${it.message}"
                                }
                                isRecoveringIce = false
                            }
                        }
                    }
                )
                started = true
            }.onFailure {
                status = it.message ?: "Failed to start consultation"
            }
        }
    }

    DisposableEffect(Unit) {
        onDispose {
            signalingClient.disconnect()
            sessionManager.dispose()
        }
    }

    LaunchedEffect(Unit) {
        if (!hasRequestedPermissions) {
            hasRequestedPermissions = true
            permissionsLauncher.launch(
                arrayOf(
                    Manifest.permission.CAMERA,
                    Manifest.permission.RECORD_AUDIO,
                ),
            )
        }
    }

    Scaffold(
        topBar = {
            Column(modifier = Modifier.padding(16.dp)) {
                Text("Consultatie Video Live")
                Text("Role: $role | Room: $roomId")
                Text("Status: $status", style = MaterialTheme.typography.bodySmall)
            }
        },
    ) { innerPadding ->
        Box(
            modifier = Modifier
                .fillMaxSize()
                .padding(innerPadding)
                .background(Color.Black),
        ) {
            AndroidView(
                factory = { viewContext ->
                    SurfaceViewRenderer(viewContext).also { renderer ->
                        remoteRenderer = renderer
                    }
                },
                modifier = Modifier.fillMaxSize(),
                update = { renderer ->
                    if (localRenderer != null && remoteRenderer != null) {
                        sessionManager.initRenderers(localRenderer!!, remoteRenderer!!)
                    }
                    remoteRenderer = renderer
                },
            )


            if (started && !hasRemoteVideo) {
                CircularProgressIndicator(modifier = Modifier.align(Alignment.Center))
            }

            AndroidView(
                factory = { viewContext ->
                    SurfaceViewRenderer(viewContext).also { renderer ->
                        localRenderer = renderer
                    }
                },
                modifier = Modifier
                    .align(Alignment.BottomEnd)
                    .padding(end = 20.dp, bottom = 110.dp)
                    .size(width = 120.dp, height = 180.dp)
                    .clip(RoundedCornerShape(12.dp)),
                update = { renderer ->
                    if (localRenderer != null && remoteRenderer != null) {
                        sessionManager.initRenderers(localRenderer!!, remoteRenderer!!)
                    }
                    localRenderer = renderer
                },
            )

            ControlBar(
                isMuted = isMuted,
                enabled = started,
                onToggleMute = {
                    val isMicEnabled = sessionManager.toggleMute()
                    isMuted = !isMicEnabled
                },
                onSwitchCamera = {
                    sessionManager.switchCamera()
                },
                onHangUp = {
                    signalingClient.disconnect()
                    sessionManager.hangUp()
                    navController.popBackStack()
                },
                modifier = Modifier
                    .align(Alignment.BottomCenter)
                    .padding(bottom = 30.dp),
            )
        }
    }
}

private suspend fun startSession(
    roomId: String,
    serverUrl: String,
    signalingClient: SocketSignalingClient,
    sessionManager: WebRtcSessionManager,
    onStatusChange: (String) -> Unit,
    onConnected: (Boolean) -> Unit,
    onRemoteVideo: () -> Unit,
    onIncomingMessage: (SignalingMessage) -> Unit,
    onReadyToOffer: () -> Unit,
    onIceFailed: () -> Unit,
) {
    Log.d(TAG, "Starting session room=$roomId server=$serverUrl")
    onStatusChange("Opening user media")
    sessionManager.openUserMedia()

    onStatusChange("Creating peer connection")
    sessionManager.createPeerConnection(
        onIceCandidate = { signalingClient.sendCandidate(roomId, it) },
        onConnectionStateChange = { state ->
            onConnected(state == PeerConnection.PeerConnectionState.CONNECTED)
            if (state == PeerConnection.PeerConnectionState.FAILED) {
                onIceFailed()
            }
            onStatusChange(
                when (state) {
                    PeerConnection.PeerConnectionState.CONNECTING -> "Connecting"
                    PeerConnection.PeerConnectionState.CONNECTED -> "Connected"
                    PeerConnection.PeerConnectionState.FAILED -> "Failed - retrying ICE"
                    PeerConnection.PeerConnectionState.DISCONNECTED -> "Disconnected"
                    PeerConnection.PeerConnectionState.CLOSED -> "Closed"
                    PeerConnection.PeerConnectionState.NEW -> "New"
                },
            )
        },
        onRemoteTrack = onRemoteVideo,
    )

    onStatusChange("Connecting signaling server")
    signalingClient.connect(
        serverUrl = serverUrl,
        roomId = roomId,
        onConnected = {
            onStatusChange("Joined room $roomId. Waiting for partner...")
        },
        onReady = {
            onStatusChange("Partner joined. Starting negotiation...")
            onReadyToOffer()
        },
        onError = { error ->
            onStatusChange("Error: $error")
        },
        onMessage = onIncomingMessage,
    )
}

private suspend fun handleSignalingMessage(
    message: SignalingMessage,
    roomId: String,
    signalingClient: SocketSignalingClient,
    sessionManager: WebRtcSessionManager,
    onStatusChange: (String) -> Unit,
) {
    runCatching {
        when (message.type) {
            "offer" -> {
                val remoteSdp = message.sdp ?: return@runCatching
                Log.d(TAG, "Received offer from peer")
                onStatusChange("Setting remote offer...")
                sessionManager.setRemoteDescription("offer", remoteSdp)
                onStatusChange("Creating answer...")
                val answer = sessionManager.createAnswer()
                signalingClient.sendAnswer(roomId, answer.description)
                onStatusChange("Answer sent")
            }

            "answer" -> {
                val remoteSdp = message.sdp ?: return@runCatching
                Log.d(TAG, "Received answer from peer")
                onStatusChange("Setting remote answer...")
                sessionManager.setRemoteDescription("answer", remoteSdp)
                onStatusChange("Connecting...")
            }

            "candidate" -> {
                val candidate = message.candidate ?: return@runCatching
                val sdpMLineIndex = message.sdpMLineIndex ?: return@runCatching
                val sdpMid = message.sdpMid ?: "0"
                Log.d(TAG, "Received ICE candidate (mid=$sdpMid, line=$sdpMLineIndex)")
                sessionManager.addCandidate(IceCandidate(sdpMid, sdpMLineIndex, candidate))
            }
        }
    }.onFailure { e ->
        Log.e(TAG, "Error handling signaling message", e)
        onStatusChange("Signaling Error: ${e.message}")
    }
}

@Composable
private fun ControlBar(
    isMuted: Boolean,
    enabled: Boolean,
    onToggleMute: () -> Unit,
    onSwitchCamera: () -> Unit,
    onHangUp: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .height(72.dp),
        horizontalArrangement = Arrangement.Center,
        verticalAlignment = Alignment.CenterVertically,
    ) {
        FloatingActionButton(
            onClick = onToggleMute,
            containerColor = Color.White,
            modifier = Modifier.padding(horizontal = 10.dp),
        ) {
            Icon(
                imageVector = if (isMuted) Icons.Default.MicOff else Icons.Default.Mic,
                contentDescription = "Mute toggle",
                tint = Color(0xFF1565C0),
            )
        }

        FloatingActionButton(
            onClick = onHangUp,
            containerColor = Color.Red,
            modifier = Modifier.padding(horizontal = 10.dp),
        ) {
            Icon(
                imageVector = Icons.Default.CallEnd,
                contentDescription = "End call",
                tint = Color.White,
            )
        }

        FloatingActionButton(
            onClick = onSwitchCamera,
            containerColor = Color.White,
            modifier = Modifier.padding(horizontal = 10.dp),
            ) {
            Icon(
                imageVector = Icons.Default.Cameraswitch,
                contentDescription = "Switch camera",
                tint = Color(0xFF1565C0),
            )
        }
    }
}

File: flutter_app/lib/src/screens/call_screen.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter_webrtc/flutter_webrtc.dart';

import '../models/signaling_message.dart';
import '../services/socket_signaling_client.dart';
import '../services/webrtc_session_manager.dart';

class CallScreen extends StatefulWidget {
  const CallScreen({
    super.key,
    required this.role,
    required this.roomId,
    required this.serverUrl,
  });

  final String role;
  final String roomId;
  final String serverUrl;

  @override
  State<CallScreen> createState() => _CallScreenState();
}

class _CallScreenState extends State<CallScreen> {
  final _localRenderer = RTCVideoRenderer();
  final _remoteRenderer = RTCVideoRenderer();
  final _sessionManager = WebRtcSessionManager();
  final _signalingClient = SocketSignalingClient();

  StreamSubscription<SignalingMessage>? _messageSub;
  bool _isMuted = false;
  bool _started = false;
  bool _offerSent = false;
  bool _isOfferer = false;
  bool _isRecoveringIce = false;
  String _status = 'Initializing call...';

  @override
  void initState() {
    super.initState();
    _init();
  }

  Future<void> _init() async {
    await _localRenderer.initialize();
    await _remoteRenderer.initialize();

    setState(() => _status = 'Opening camera and microphone...');
    final local = await _sessionManager.openUserMedia();
    _localRenderer.srcObject = local;

    setState(() => _status = 'Creating peer connection...');
    await _sessionManager.createPeerConnection(
      onIceCandidate: (candidate) {
        _signalingClient.sendCandidate(
          widget.roomId,
          candidate.candidate ?? '',
          candidate.sdpMid,
          candidate.sdpMLineIndex,
        );
      },
      onConnectionStateChange: (state) async {
        if (state == RTCPeerConnectionState.RTCPeerConnectionStateFailed) {
          if (_started && _isOfferer && !_isRecoveringIce) {
            _isRecoveringIce = true;
            setState(() => _status = 'Connection failed, restarting ICE...');
            try {
              final restartOffer = await _sessionManager.createIceRestartOffer();
              _signalingClient.sendOffer(widget.roomId, restartOffer.sdp ?? '');
              setState(() => _status = 'ICE restart offer sent');
            } catch (e) {
              setState(() => _status = 'ICE restart error: $e');
            }
            _isRecoveringIce = false;
          } else {
            setState(() => _status = 'Connection failed');
          }
          return;
        }
        setState(() => _status = _mapConnectionState(state));
      },
      onRemoteStream: (remote) {
        setState(() => _remoteRenderer.srcObject = remote);
      },
    );

    _messageSub = _signalingClient.messages.listen((message) async {
      await _handleSignalingMessage(message);
    });

    setState(() => _status = 'Connecting signaling server...');
    await _signalingClient.connect(
      serverUrl: widget.serverUrl,
      roomId: widget.roomId,
      onConnected: () => setState(() => _status = 'Joined room. Waiting for partner...'),
      onReady: () async {
        if (_offerSent) return;
        _offerSent = true;
        _isOfferer = true;
        setState(() => _status = 'Creating offer...');
        try {
          final offer = await _sessionManager.createOffer();
          _signalingClient.sendOffer(widget.roomId, offer.sdp ?? '');
          setState(() {
            _status = 'Offer sent, waiting for answer';
            _started = true;
          });
        } catch (e) {
          _offerSent = false;
          setState(() => _status = 'Offer error: $e');
        }
      },
      onError: (error) => setState(() => _status = error),
    );
  }

  String _mapConnectionState(RTCPeerConnectionState state) {
    switch (state) {
      case RTCPeerConnectionState.RTCPeerConnectionStateConnecting:
        return 'Connecting';
      case RTCPeerConnectionState.RTCPeerConnectionStateConnected:
        return 'Connected';
      case RTCPeerConnectionState.RTCPeerConnectionStateDisconnected:
        return 'Disconnected';
      case RTCPeerConnectionState.RTCPeerConnectionStateClosed:
        return 'Closed';
      case RTCPeerConnectionState.RTCPeerConnectionStateNew:
        return 'New';
      case RTCPeerConnectionState.RTCPeerConnectionStateFailed:
        return 'Failed';
    }
  }

  Future<void> _handleSignalingMessage(SignalingMessage message) async {
    try {
      switch (message.type) {
        case 'offer':
          final remoteSdp = message.sdp;
          if (remoteSdp == null) return;
          setState(() => _status = 'Setting remote offer...');
          await _sessionManager.setRemoteDescription('offer', remoteSdp);
          setState(() => _status = 'Creating answer...');
          final answer = await _sessionManager.createAnswer();
          _signalingClient.sendAnswer(widget.roomId, answer.sdp ?? '');
          setState(() {
            _status = 'Answer sent';
            _started = true;
          });
          break;
        case 'answer':
          final remoteSdp = message.sdp;
          if (remoteSdp == null) return;
          setState(() => _status = 'Setting remote answer...');
          await _sessionManager.setRemoteDescription('answer', remoteSdp);
          setState(() => _status = 'Connecting...');
          break;
        case 'candidate':
          if (message.candidate == null || message.sdpMLineIndex == null) return;
          await _sessionManager.addCandidate(
            RTCIceCandidate(
              message.candidate!,
              message.sdpMid,
              message.sdpMLineIndex!,
            ),
          );
          break;
      }
    } catch (e) {
      setState(() => _status = 'Signaling error: $e');
    }
  }

  @override
  void dispose() {
    _messageSub?.cancel();
    _signalingClient.dispose();
    _sessionManager.dispose();
    _localRenderer.dispose();
    _remoteRenderer.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Video Consultation | ${widget.role} | ${widget.roomId}'),
      ),
      body: Stack(
        children: [
          Positioned.fill(
            child: Container(
              color: Colors.black,
              child: RTCVideoView(
                _remoteRenderer,
                objectFit: RTCVideoViewObjectFit.RTCVideoViewObjectFitCover,
              ),
            ),
          ),
          Positioned(
            right: 16,
            bottom: 120,
            width: 120,
            height: 180,
            child: ClipRRect(
              borderRadius: BorderRadius.circular(12),
              child: RTCVideoView(
                _localRenderer,
                mirror: true,
                objectFit: RTCVideoViewObjectFit.RTCVideoViewObjectFitCover,
              ),
            ),
          ),
          Positioned(
            top: 12,
            left: 12,
            right: 12,
            child: Container(
              padding: const EdgeInsets.all(12),
              color: Colors.black54,
              child: Text(
                'Status: $_status',
                style: const TextStyle(color: Colors.white),
              ),
            ),
          ),
          Align(
            alignment: Alignment.bottomCenter,
            child: Padding(
              padding: const EdgeInsets.only(bottom: 24),
              child: Row(
                mainAxisAlignment: MainAxisAlignment.center,
                children: [
                  FloatingActionButton(
                    heroTag: 'mute',
                    onPressed: () {
                      final micEnabled = _sessionManager.toggleMute();
                      setState(() => _isMuted = !micEnabled);
                    },
                    child: Icon(_isMuted ? Icons.mic_off : Icons.mic),
                  ),
                  const SizedBox(width: 16),
                  FloatingActionButton(
                    heroTag: 'hangup',
                    backgroundColor: Colors.red,
                    onPressed: () async {
                      await _sessionManager.hangUp();
                      if (mounted) Navigator.of(context).pop();
                    },
                    child: const Icon(Icons.call_end),
                  ),
                  const SizedBox(width: 16),
                  FloatingActionButton(
                    heroTag: 'switch',
                    onPressed: () => _sessionManager.switchCamera(),
                    child: const Icon(Icons.cameraswitch),
                  ),
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }
}
